server/backends/disk_cache/disk_cache.go

```diff
@@ -101,35 +101,48 @@ func (c *DiskCache) WithPrefix(prefix string) interfaces.Cache {
 	}
 
 	return &DiskCache{
-		l:       c.l,
-		rootDir: c.rootDir,
-		lock:    c.lock,
-		prefix:  newPrefix,
+		l:            c.l,
+		rootDir:      c.rootDir,
+		lock:         c.lock,
+		prefix:       newPrefix,
+		diskIsMapped: c.diskIsMapped,
 	}
 }
 
 func (c *DiskCache) initializeCache() error {
 	if err := disk.EnsureDirectoryExists(c.rootDir); err != nil {
 		return err
 	}
-	records := make([]*fileRecord, 0)
-	walkFn := func(path string, info os.FileInfo, err error) error {
-		if !info.IsDir() {
-			records = append(records, makeRecord(path, info))
+	c.diskIsMapped = false
+	go func() {
+		start := time.Now()
+		records := make([]*fileRecord, 0)
+		walkFn := func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !info.IsDir() {
+				records = append(records, makeRecord(path, info))
+			}
+			return nil
+		}
+		if err := filepath.Walk(c.rootDir, walkFn); err != nil {
+			log.Warningf("Error walking disk directory: %s", err)
 		}
-		return nil
-	}
-	if err := filepath.Walk(c.rootDir, walkFn); err != nil {
-		return err
-	}
 
-	// Sort entries by ascending ATime.
-	sort.Slice(records, func(i, j int) bool { return records[i].lastUse.Before(records[j].lastUse) })
+		// Sort entries by ascending ATime.
+		sort.Slice(records, func(i, j int) bool { return records[i].lastUse.Before(records[j].lastUse) })
 
-	// Populate our state tracking datastructures.
-	for _, record := range records {
-		c.l.Add(record.key, record)
-	}
+		// Populate our state tracking datastructures.
+		for _, record := range records {
+			c.lock.Lock()
+			c.l.Add(record.key, record)
```

Comment:
(assuming the lru is nearly full) will adding all of these records to the front of the LRU make it so that files accessed while the disk is being mapped are likely to get evicted soon after all of these records are added?

ideally we'd add them after the records accessed during mapping or something like that

Reply:
Yeah I see what you're saying and that could happen. I think adding anything we add during the loading phase to that list of records we sort and add to the LRU should fix this.

Reply:
This turned out to be a little difficult. The only way to guarantee these files are always visible (thinking of builds without the bytes here) is by locking the LRU structure while we batch-add everything to it.

We obviously don't lock during the file-scanning part, which is the slow bit.